### PR TITLE
Support parent POMs with version range

### PR DESCRIPTION
**Pull request by blindpirate**
See #1898 for more details.

As stated in [Maven 3.2.2 release notes](https://maven.apache.org/docs/3.2.2/release-notes.html) ,
Maven supports parent POM version with range. When Gradle resolves
parent POMs, it didn't take version range into consideration. This PR
adds some resolution for dynamic parent POM version.

Since Gradle claims 100% compatibility with Maven, I think this improvement is necessary.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [x] [Sign Gradle CLA](http://gradle.org/contributor-license-agreement/)
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### CODE REVIEW

**Comment by bmuschko** [comment]
@blindpirate Can you please find a reviewer for your PR?

**Comment by blindpirate** [comment]
Sure.

@bigdaz Could you please review it for me? I saw you were in the list of "recently edited these files".

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolverDescriptorParseContext.java [hunk 1]
```diff
@@ -49,10 +60,25 @@ public ExternalResourceResolverDescriptorParseContext(ComponentResolvers mainRes
...
+        moduleComponentIdentifier = resolveDynamicVersionIfNecessary(moduleComponentIdentifier);
         File resolvedArtifactFile = resolveMetaDataArtifactFile(moduleComponentIdentifier, mainResolvers.getComponentResolver(), mainResolvers.getArtifactResolver(), artifactType);
         return fileResourceRepository.resource(resolvedArtifactFile);
     }
 
+    private ModuleComponentIdentifier resolveDynamicVersionIfNecessary(ModuleComponentIdentifier moduleComponentIdentifier) {
+        ModuleVersionSelector selector = DefaultModuleVersionSelector.newSelector(moduleComponentIdentifier.getGroup(),
+            moduleComponentIdentifier.getModule(),
+            moduleComponentIdentifier.getVersion());
+        DependencyMetadata metadata = new MavenDependencyMetadata(MavenScope.Compile, false, selector, new ArrayList<Artifact>(), new ArrayList<Exclude>());
```
Comment (bigdaz) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolverDescriptorParseContext.java]: This type `(ExternalResourceResolverDescriptorParseContext`) should not be tied to Maven in this way. The simplest option is probably to have  the caller supply a `DependencyMetadata` instance instead of a `ModuleComponentSelector` (when we need to support dynamic versions). 

So my previous suggestion becomes:
```
public interface DescriptorParseContext {
    LocallyAvailableExternalResource getMetaDataArtifact(ModuleComponentIdentifier componentIdentifier, ArtifactType artifactType);
    LocallyAvailableExternalResource getMetaDataArtifact(DependencyMetadata dependency, ArtifactType artifactType);
}
```

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolverDescriptorParseContext.java [hunk 2]
```diff
@@ -49,10 +60,25 @@ public ExternalResourceResolverDescriptorParseContext(ComponentResolvers mainRes
...
+        BuildableComponentIdResolveResult idResolveResult = new DefaultBuildableComponentIdResolveResult();
+        mainResolvers.getComponentIdResolver().resolve(metadata, idResolveResult);
+
+        moduleComponentIdentifier = new DefaultModuleComponentIdentifier(moduleComponentIdentifier.getGroup(),
+            moduleComponentIdentifier.getModule(),
+            idResolveResult.getModuleVersionId().getVersion());
```
Comment (bigdaz) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolverDescriptorParseContext.java]: In general, prefer `[Module|Project]Component[Identifier|Selector]` over `ModuleVersion[Identifier|Selector]`. 

So here you can just do: `idResolveResult.getId()` and use the returned `ComponentIdentifier` as input into the next stage of resolution.

A _less good_ alternative is:  `DefaultModuleComponentIdentifier.newId(idResolveResult.getModuleVersionId())`

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenVersionRangeResolveIntegrationTest.groovy [hunk 3]
```diff
@@ -58,4 +58,43 @@ dependencies {
...
+configurations { compile }
+
+dependencies {
+    compile group: "org.test", name: "child", version: "1.0"
+}
+"""
+        and:
+        mavenRepo.module('org.test', 'child', '1.0').parent('org.test', 'parent', '[2.0,3.0)').publish()
+        mavenRepo.module('org.test', 'parent', '2.0').dependsOn('org.test', 'dep', '2.0').publishPom()
+        mavenRepo.module('org.test', 'parent', '2.1').dependsOn('org.test', 'dep', '2.1').publishPom()
```
Comment (bigdaz) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenVersionRangeResolveIntegrationTest.groovy]: For completeness, I'd publish a version `3.0` here just to demonstrate that the range is being fully respected.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenVersionRangeResolveIntegrationTest.groovy [hunk 4]
```diff
@@ -58,4 +58,43 @@ dependencies {
             }
         }
     }
+
+    @Issue("https://github.com/gradle/gradle/issues/1898")
+    def "can resolve parent pom with version range"() {
```
Comment (bigdaz) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenVersionRangeResolveIntegrationTest.groovy]: I wonder if it's also possible to have a version range in the coordinates for an _imported_ POM?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolverDescriptorParseContext.java (hunk 1 above)
Comment (blindpirate) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolverDescriptorParseContext.java]: Done. Please take a look at it again.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolverDescriptorParseContext.java (hunk 2 above)
Comment (blindpirate) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolverDescriptorParseContext.java]: Done. See [here](https://github.com/blindpirate/gradle/blob/464c412b272e0a64bd25d1d9fc0ab71144ec617f/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolverDescriptorParseContext.java#L64)

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenVersionRangeResolveIntegrationTest.groovy (hunk 4 above)
Comment (blindpirate) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenVersionRangeResolveIntegrationTest.groovy]: [Here](https://github.com/blindpirate/gradle/blob/464c412b272e0a64bd25d1d9fc0ab71144ec617f/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenVersionRangeResolveIntegrationTest.groovy#L104) it is.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenVersionRangeResolveIntegrationTest.groovy (hunk 3 above)
Comment (blindpirate) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenVersionRangeResolveIntegrationTest.groovy]: Sure. I added `1.0` and `3.0`

**Comment by blindpirate** [comment]
@bigdaz A little improvement after your constructive advice.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/GradlePomModuleDescriptorParser.java [hunk 5]
```diff
@@ -204,10 +209,20 @@ private PomReader parseParentPom(DescriptorParseContext parseContext, ModuleComp
...
-        LocallyAvailableExternalResource localResource = parseContext.getMetaDataArtifact(parentId, ArtifactType.MAVEN_POM);
+        LocallyAvailableExternalResource localResource = getMetaDataArtifact(parseContext, parentId);
         PomReader pomReader = new PomReader(localResource, moduleIdentifierFactory, childProperties);
         GradlePomModuleDescriptorBuilder mdBuilder = new GradlePomModuleDescriptorBuilder(pomReader, gradleVersionSelectorScheme, mavenVersionSelectorScheme, moduleIdentifierFactory, moduleExclusions);
         doParsePom(parseContext, mdBuilder, pomReader);
         return pomReader;
     }
+
+    private LocallyAvailableExternalResource getMetaDataArtifact(DescriptorParseContext parseContext, ModuleComponentIdentifier parentId) {
+        if (mavenVersionSelectorScheme.parseSelector(parentId.getVersion()).isDynamic()) {
```
Comment (bigdaz) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/GradlePomModuleDescriptorParser.java]: We shouldn't be doing this. If the input is a _selector_, meaning it _might_ have a dynamic version, then we should treat it as such. Don't even check if the version looks like it's dynamic. We either have a `ComponentIdentifier` (in which case we know exactly the component to use, or a `ComponentSelector` in which case we do some logic to get the identifier. Part of this logic is checking for a dynamic version: but this is done inside the dependency resolution engine.

In summary: if the input is a selector (might have a dynamic version), then we should treat it like it is dynamic and let the dependency resolution engine to the conversion.

**Comment by bigdaz** [comment]
Sorry @blindpirate, I won't have time to do any further review before my vacation. I've asked @big-guy to take over the review for now.

I would still prefer that if an input is a _selector_ we always put it through dependency resolution as such (I left a comment).
I think this means that [here](https://github.com/gradle/gradle/blob/73dc402362cc284ab08940a1909f084208c5effc/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/GradlePomModuleDescriptorParser.java#L97-L100) we should be always creating a `ModuleComponentSelector`.

**Comment by blindpirate** [comment]
@bigdaz would you like to take a look at it when you're free?

**Comment by bigdaz** [comment]
@blindpirate Can you please create a GH issue for this fix, so it will be included in the release notes?

**Comment by blindpirate** [comment]
@bigdaz Sure.
